ClientAwakeTimeProvider API to reduce the 93 sec default value

carlosgp143 opened the pull request:
Hi! 

After the first version of the Queue Mode, here is a new PR to add the feature of reducing the 93 seconds of client awake time. The interface `ClientAwakeTimeInformation` is created, which has the method `usedClientAwakeTime(reg)` for giving the awake time for an specific client. If the user don't do anything, the `LeshanServerBuilder` uses the `ClientAwakeTimeInformationDefaultImpl` which returns 93 sec for all the clients. The user can also add a different `ClientAwakeTimeInformation` object using the method `setClientAwakeTimeInfo` in the `LeshanServerBuilder`. For just changing the 93 seconds, a `ClientAwakeTimeInformationDefaultImpl` object can be created with a different time as a parameter in the constructor.

I've also implemented a Dynamic way to select this time, in the `ClientAwakeTimeInformationDynamicImpl`, which reads the `AdditionalRegistrationAttributes` in order to select the timer. One thing I've found in this implementation is that these additional parameters does not seem to be updated when an update message arrives... Or does it and I've miss it somehow? 
This means that the first value that is sent in the registration message is the one included all the time. I've trying to go a bit through the code and the `update` method seems to just take the same Map all the time, without uploading it. So coming back to Queue Mode, I think the client should be able to change the awake time at any moment and send it in the update message... so I think we should do something to update the `AdditionalRegistrationAttributes`.

And btw, I don't know if the new features that need to be added should be introduced in the queue_mode branch or directly in master (I selected master as base branch now but feel free to change it).

All feedback is appreciated :-)

sbernard31 commented on leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java:
```diff
@@ -67,7 +72,7 @@ public boolean isClientAwake(Registration registration) {
     public void setAwake(Registration reg) {
         if (reg.usesQueueMode()) {
-            PresenceStatus status = new PresenceStatus();
+            PresenceStatus status = new PresenceStatus(awakeTimeInfo.usedClientAwakeTime(reg));
```
I'm even asking myself if we can not consider that a device is `awake` when it is in the `clientStatusList` of `PresenceServiceImpl` and `sleeping` when it is not.

If we do that, I feel we don't need `PresenceStatus` anymore as the only remain attribute is `clientScheduledFuture` so we can put it directly in the `clientStatusList` map.

carlosgp143 replied:
Yes, it's true that `clientAwakeTime` is not needed inside `PresenceStatus` anymore, since it is directly taken with the Awake Time Provider by `PresenceService`. 

> I'm even asking myself if we can not consider that a device is awake when it is in the clientStatusList of PresenceServiceImpl and sleeping when it is not.

I think that is simplifying it too much and making the code more unclear... `setSleeping` will be to remove a `ScheduledFuture` from a Map, and `setAwake` to add it... Doesn't sound very intuitive FMPOV. The enumerate with `Awake` and `Sleeping` states makes it more clear I think. Also in my original implementation, the `PresenceStatus` was a way to handle every client's queue mode independently. For example, the queues to store requests are also inside this object... We can do it your way but I think is more clear and can be more useful in a future to have this `PresenceStatus` object. Otherwise adding more features could be difficult, and of course we can also remove it in the future if we still see that there is no point in having this object. 

What do you think?

sbernard31 replied:
>  For example, the queues to store requests are also inside this object...

FMPOV, If we implements this "queue of requests", it  should be handled by another object which just use the PresenceService and Send request. I mean this is on top of what we are implementing not inside.

I also feel this could simplify thread-safeness, but let's this aside. I will maybe try it on my side and propose a PR to talk about this specifically. Not critical this is just internal implementation.

carlosgp143 replied:
Ok, I also don't think is a critical aspect right now, we can discuss it in another PR.

carlosgp143 commented on leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/ClientAwakeTimeInformationDynamicImpl.java:
```diff
@@ -0,0 +1,44 @@
+
+import org.eclipse.leshan.server.registration.Registration;
+
+public class ClientAwakeTimeInformationDynamicImpl implements ClientAwakeTimeInformation {
```
It is out of the spec but the same as the default provider with a different time than 93 seconds... We can do it that way but I think it doesn't hurt to have it directly in the Code base... Remember that the 93 seconds are only a recommendation in the specification.

sbernard31 replied:
> it doesn't hurt to have it directly in the Code base

It's right but I also feel that keeping small the scope of a project is a good thing. It's hard to find the good balance.

carlosgp143 replied:
So... not including this class and make a wiki page explaining how to develop this class is what you would do?

sbernard31 replied:
Personally, I would do that. @msangoi do you have any opinion about that ?

carlosgp143 commented:
Hi! Thanks for the feedback. I agree in the naming, I'll make the necessary changes. 

About the update for additional registration parameters, I have taken a look and it needs several modifications... `RegisterResource` needs changes in the `handleUpdate` method. `UpdateRequest` needs the `additionalRegistrationAttributes` as an attribute, `RegistrationUpdate` needs also changes in the `update` method, and also `RegistrationHandler` in the `update` method. This is pretty big, and not strictly necesary for this PR, it is only better for the dynamic provider, but it could work without it (the client only gives the awake time in the registration message). Should I do a different PR for updating the `additionalRegistrationAttributes` ?

If we agree on that, then this PR only needs naming changes and reach an agreement in what we should do with the `PresenceStatus`... So I'll wait for your feedback :-)

sbernard31 commented:
About the update, if you don't need it. We don't need to implement it.
If you need it, I prefer to go with a separated request.
If we go for it, we need to specify behavior of updating `additionalRegistrationAttributes`, I mean "Does it replace previous one, does it complete, is it a problem if we can not removing it ? ...

carlosgp143 commented:
I think the update is necessary if we want to have a real dynamic way to announce the client awake time. Otherwise, it can just set it in the registration message, but it cannot change at any time. I think is a much better implementation making the client able to change the awake time through update messages. With this the client can adapt better to different kind of situations. 

I guess the update behaviour should be to check if the attribute is already included in the `additionalRegistrationAttributes`, and if that is the case change the value to the new one. If the attribute is not present, it should just add this attribute. All the other attributes present in the Map remain the same. And of course this should be done with all the attributes that are included in the update message.  

I think I can make a separate PR implementing that, it shouldn't take much effort.

carlosgp143 commented:
Hi! I made all the changes in the naming, removed the `clientAwakeTime` from the `PresenceStatus`, and also removed completely the class `DynamicClientAwakeTimeProvider`. Anything else to be done for this PR? How do we proceed with the wiki page about the `DynamicClientAwakeTimeProvider`?

I'll also move on to the PR for implementing the update of the `additionalRegistrationAttributes`. This should be done to master branch since is not really a queue mode feature, or also to the queue_mode one?

sbernard31 commented on leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/DefaultClientAwakeTimeProvider.java:
```diff
@@ -17,20 +17,20 @@
 import org.eclipse.leshan.server.registration.Registration;
 
-public class ClientAwakeTimeInformationDefaultImpl implements ClientAwakeTimeInformation {
+public class DefaultClientAwakeTimeProvider implements ClientAwakeTimeProvider {
```
Finally, I find `StaticClientAwakeTimeProvider` more appropriate.
If you are OK I would change it before to integrate it in master.

carlosgp143 replied:
Sure! :-)

sbernard31 commented on leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java:
```diff
@@ -91,7 +91,7 @@
     private boolean noQueueMode = false;
 
-    private ClientAwakeTimeInformation awakeTimeInfo = new ClientAwakeTimeInformationDefaultImpl();
+    private ClientAwakeTimeProvider awakeTimeProvider = new DefaultClientAwakeTimeProvider();
```
I prefer to create the object only if needed in build method.

carlosgp143 commented:
> I can implement those 2 remarks and merge this PR in master if you are OK.

I have no objections :-)

> I create a page you can complete it. (You can rename it too)
Once this will be finished we could add a link in the wiki right sidebar

I'll take a look into it tomorrow and modify it.

> About update, you could go for a new PR targeting master.

Prefect! I've already been implementing it so it will come soon.

sbernard31 commented:
Integrated in master (commit 1c322bf6efc077c433bc2cfbb6324cd2abce9571)
I made the 2 changes we talk about and severals minor clean-up.
